File: _includes/sort-benchmark/algorithms/van_emde_boas.rs
struct VebTree {
    universe: usize,
    min: Option<usize>,
    max: Option<usize>,
    summary: Option<Box<VebTree>>,
    cluster: Vec<Option<Box<VebTree>>>,
}

fn lower_sqrt(universe: usize) -> usize {
    1usize << (universe.trailing_zeros() / 2)
}

impl VebTree {
    fn new(universe: usize) -> Self {
        debug_assert!(universe.is_power_of_two());
        debug_assert!(universe >= 2);

        if universe == 2 {
            return Self {
                universe,
                min: None,
                max: None,
                summary: None,
                cluster: Vec::new(),
            };
        }

        let lower = lower_sqrt(universe);
        let upper = universe / lower;

        Self {
            universe,
            min: None,
            max: None,
            summary: Some(Box::new(Self::new(upper))),
            cluster: (0..upper).map(|_| None).collect(),
        }
    }

    fn high(&self, x: usize) -> usize {
        x / lower_sqrt(self.universe)
    }

    fn low(&self, x: usize) -> usize {
        x % lower_sqrt(self.universe)
    }

    fn index(&self, high: usize, low: usize) -> usize {
        high * lower_sqrt(self.universe) + low
    }

    fn minimum(&self) -> Option<usize> {
        self.min
    }

    fn maximum(&self) -> Option<usize> {
        self.max
    }

    fn cluster_mut(&mut self, i: usize) -> &mut VebTree {
        let lower = lower_sqrt(self.universe);
        self.cluster[i].get_or_insert_with(|| Box::new(Self::new(lower)))
    }

    fn empty_insert(&mut self, x: usize) {
        self.min = Some(x);
        self.max = Some(x);
    }

    fn insert(&mut self, mut x: usize) {
        if self.min.is_none() {
            self.empty_insert(x);
            return;
        }

        if x < self.min.unwrap() {
            let old_min = self.min.unwrap();
            self.min = Some(x);
            x = old_min;
        }

        if self.universe > 2 {
            let h = self.high(x);
            let l = self.low(x);
            if self.cluster[h]
                .as_ref()
                .and_then(|c| c.minimum())
                .is_none()
            {
                self.summary.as_mut().unwrap().insert(h);
                self.cluster_mut(h).empty_insert(l);
            } else {
                self.cluster_mut(h).insert(l);
            }
        }

        if x > self.max.unwrap() {
            self.max = Some(x);
        }
    }

    fn successor(&self, x: usize) -> Option<usize> {
        if self.universe == 2 {
            return if x == 0 && self.max == Some(1) {
                Some(1)
            } else {
                None
            };
        }

        if let Some(min) = self.min {
            if x < min {
                return Some(min);
            }
        } else {
            return None;
        }

        let h = self.high(x);
        let l = self.low(x);
        let max_low = self.cluster[h].as_ref().and_then(|c| c.maximum());
        if max_low.is_some_and(|m| l < m) {
            let offset = self.cluster[h].as_ref().unwrap().successor(l).unwrap();
            return Some(self.index(h, offset));
        }

        let succ_cluster = self.summary.as_ref().unwrap().successor(h)?;
        let offset = self.cluster[succ_cluster]
            .as_ref()
            .and_then(|c| c.minimum())
            .unwrap();
        Some(self.index(succ_cluster, offset))
    }
}

fn van_emde_boas_sort(a: &mut [usize]) {
    if a.len() <= 1 {
        return;
    }

    let min = *a.iter().min().unwrap();
    let max = *a.iter().max().unwrap();
    let span = max - min + 1;
    let mut count = vec![0usize; span];

    for &x in a.iter() {
        count[x - min] += 1;
    }

    let universe = span.next_power_of_two().max(2);
    let mut tree = VebTree::new(universe);

    for (offset, &c) in count.iter().enumerate() {
        if c > 0 {
            tree.insert(offset);
        }
    }

    let mut idx = 0;
    let mut cur = tree.minimum();
    while let Some(v) = cur {
        let value = min + v;
        for _ in 0..count[v] {
            a[idx] = value;
            idx += 1;
        }
        cur = tree.successor(v);
    }
}

File: _includes/sort-benchmark/algorithms/van_emde_boas.swift
final class VebTree {
    var universe: Int
    var min: Int?
    var max: Int?
    var summary: VebTree?
    var cluster: [VebTree?]

    init(universe: Int) {
        assert(universe > 0 && (universe & (universe - 1)) == 0)
        assert(universe >= 2)

        self.universe = universe
        self.min = nil
        self.max = nil

        if universe == 2 {
            self.summary = nil
            self.cluster = []
            return
        }

        let lower = VebTree.lower_sqrt(universe)
        let upper = universe / lower

        self.summary = VebTree(universe: upper)
        self.cluster = Array(repeating: nil, count: upper)
    }

    static func lower_sqrt(_ universe: Int) -> Int {
        1 << (universe.trailingZeroBitCount / 2)
    }

    func high(_ x: Int) -> Int {
        x / VebTree.lower_sqrt(universe)
    }

    func low(_ x: Int) -> Int {
        x % VebTree.lower_sqrt(universe)
    }

    func index(_ high: Int, _ low: Int) -> Int {
        high * VebTree.lower_sqrt(universe) + low
    }

    func minimum() -> Int? {
        min
    }

    func maximum() -> Int? {
        max
    }

    func cluster_mut(_ i: Int) -> VebTree {
        let lower = VebTree.lower_sqrt(universe)
        if cluster[i] == nil {
            cluster[i] = VebTree(universe: lower)
        }
        return cluster[i]!
    }

    func empty_insert(_ x: Int) {
        min = x
        max = x
    }

    func insert(_ x: Int) {
        var x = x
        if min == nil {
            empty_insert(x)
            return
        }

        if x < min! {
            let old_min = min!
            min = x
            x = old_min
        }

        if universe > 2 {
            let h = high(x)
            let l = low(x)
            if cluster[h]?.minimum() == nil {
                summary!.insert(h)
                cluster_mut(h).empty_insert(l)
            } else {
                cluster_mut(h).insert(l)
            }
        }

        if x > max! {
            max = x
        }
    }

    func successor(_ x: Int) -> Int? {
        if universe == 2 {
            if x == 0 && max == 1 {
                return 1
            } else {
                return nil
            }
        }

        if let min = min {
            if x < min {
                return min
            }
        } else {
            return nil
        }

        let h = high(x)
        let l = low(x)
        let max_low = cluster[h]?.maximum()
        if let m = max_low, l < m {
            let offset = cluster[h]!.successor(l)!
            return index(h, offset)
        }

        guard let succ_cluster = summary!.successor(h) else {
            return nil
        }
        let offset = cluster[succ_cluster]!.minimum()!
        return index(succ_cluster, offset)
    }
}

func next_power_of_two(_ n: Int) -> Int {
    if n <= 1 {
        return 1
    }
    var v = UInt(bitPattern: n - 1)
    v |= v >> 1
    v |= v >> 2
    v |= v >> 4
    v |= v >> 8
    v |= v >> 16
    v |= v >> 32
    return Int(bitPattern: v &+ 1)
}

func van_emde_boas_sort(_ a: inout [Int]) {
    a.withUnsafeMutableBufferPointer { van_emde_boas_sort($0) }
}

func van_emde_boas_sort(_ a: UnsafeMutableBufferPointer<Int>) {
    if a.count <= 1 {
        return
    }

    var min = a[0]
    var max = a[0]
    for i in 1..<a.count {
        if a[i] < min { min = a[i] }
        if a[i] > max { max = a[i] }
    }
    let span = max - min + 1
    var count = [Int](repeating: 0, count: span)

    for i in 0..<a.count {
        count[a[i] - min] += 1
    }

    let universe = Swift.max(next_power_of_two(span), 2)
    let tree = VebTree(universe: universe)

    for offset in 0..<count.count {
        if count[offset] > 0 {
            tree.insert(offset)
        }
    }

    var idx = 0
    var cur = tree.minimum()
    while let v = cur {
        let value = min + v
        for _ in 0..<count[v] {
            a[idx] = value
            idx += 1
        }
        cur = tree.successor(v)
    }
}
